### PR TITLE
Made the linter print out more information in its output and fixed bugs

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

* Gave more context to the linter's output.  It was previously confusing, especially for people who don't contribute often to the engine.
* fixed the logic for finding diff files
* gave files the option to opt out of linting with the `// FLUTTER_NOLINT` comment
* started ignoring third_party code

## Related Issues

https://github.com/flutter/flutter/issues/61875

## Tests

I added the following tests:

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by gaaclarke** [comment]
@zanderso I implemented the "allow list" approach we talked about, but instead of making a monolithic list i rely on a comment in the source code to ignore linting.

File: ci/lint.dart [hunk 1]
```diff
@@ -89,6 +102,25 @@ Future<List<String>> dirContents(String repoPath) {
...
+    return true;
+  } else {
+    final RegExp exp = RegExp(r'//.*FLUTTER_NOLINT');
+    await for (String line in File(path.substring(6))
+        .openRead()
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())) {
+      if (exp.hasMatch(line)) {
+        return true;
+      } else if (line.length > 0 && line[0] != '\n' && line[0] != '/') {
```
Comment (zanderso) [inline ci/lint.dart:116]: Could you add a comment here about what this case is doing?

File: ci/lint.dart [hunk 2]
```diff
@@ -108,20 +140,25 @@ void main(List<String> arguments) async {
...
-    process.stdout.transform(utf8.decoder).listen((data) {
-      print(data);
-      exitCode = 1;
-    });
-    await process.exitCode;
+    if (!(await shouldIgnoreFile(command.file))) {
+      final String tidyArgs = calcTidyArgs(command);
+      final List<String> args = [command.file, checks, '--'];
+      args.addAll(tidyArgs.split(' '));
+      print('🔶 linting ${command.file}');
```
Comment (zanderso) [inline ci/lint.dart:151]: Maybe not for this PR, but consider adding a --verbose flag, and giving output only if there is an error when --verbose is not passed.

File: ci/lint.dart (hunk 1 above)
Comment (gaaclarke) [reply ci/lint.dart:116]: Done.

File: ci/lint.dart (hunk 2 above)
Comment (gaaclarke) [reply ci/lint.dart:155]: Good call, added comment.  I don't want to do it now until we've chased out all the problems with finding the diff'ed files.

**Pull request merged**